### PR TITLE
Improving validation for disk configurations

### PR DESCRIPTION
**Pull request by MaxBrodeurUrbas**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/tools/cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [ ] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

Adding checks in the isValid method for disk objects.
- Checking the MaxSize defined is valid. If MaxSize is set to 0 when no target disk is specified the disk isnt zeroed properly. If MaxSize is smaller than the partitions, build will fail. 
- Checking that partitions don't overlap in the config.  
- Added unit tests for these checks

###### Does this affect the toolchain?  

**NO**

###### Test Methodology

- Manual edge case testing and local builds

### CODE REVIEW

File: toolkit/tools/imagegen/configuration/disk.go [hunk 1]
```diff
@@ -21,13 +22,72 @@ type Disk struct {
...
+		return partIntervals[i][0] < partIntervals[j][0]
+	})
+	//confirm each partition ends before the next starts
+	for i := 0; i < len(partIntervals); i++ {
+		if i < len(partIntervals)-1 {
+			if partIntervals[i][1] > partIntervals[i+1][0] {
+				return fmt.Errorf("a [Partition] with an end location %d overlaps with a [Partition] with a start location %d", partIntervals[i][1], partIntervals[i+1][0])
+			}
+		}
+	}
```
Comment (PawelWMS) [inline toolkit/tools/imagegen/configuration/disk.go:43]:
```suggestion
	for i := 0; i < len(partIntervals)-1; i++ {
		if partIntervals[i][1] > partIntervals[i+1][0] {
			return fmt.Errorf("a [Partition] with an end location %d overlaps with a [Partition] with a start location %d", partIntervals[i][1], partIntervals[i+1][0])
		}
	}
```

File: toolkit/tools/imagegen/configuration/disk.go [hunk 2]
```diff
@@ -21,13 +22,72 @@ type Disk struct {
...
+// also confirms that the MaxSize defined is large enough to accomodate all partitions. No partition should have an
+// end position that exceeds the MaxSize
+func checkMaxSizeCorrectness(disk *Disk) (err error) {
+	const (
+		realDiskType = "path"
+	)
+	//MaxSize is not relevant if target disk is specified.
+	if disk.TargetDisk.Type != realDiskType {
+		if disk.MaxSize <= 0 && len(disk.Partitions) != 0 {
+			return fmt.Errorf("a configuration without a defined target disk must have a non-zero MaxSize")
```
Comment (PawelWMS) [inline toolkit/tools/imagegen/configuration/disk.go:57]: I'd also mention something about the `len(disk.Partitions) != 0` check. If I understand this check correctly, it is fine to not have a positive `MaxSize` if we don't define any partitions.

File: toolkit/tools/imagegen/configuration/disk.go [hunk 3]
```diff
@@ -21,13 +22,72 @@ type Disk struct {
...
+		//check last parition end location does not surpass MaxSize
+		for _, part := range disk.Partitions {
+			if part.End == 0 {
+				lastPartitionEnd = part.Start
+			} else if part.End > lastPartitionEnd {
+				lastPartitionEnd = part.End
+			}
+		}
+		if maxSize < lastPartitionEnd {
+			return fmt.Errorf("the MaxSize of %d is not large enough to accomodate defined partitions ending at %d.", int(maxSize), int(lastPartitionEnd))
```
Comment (PawelWMS) [inline toolkit/tools/imagegen/configuration/disk.go:68]: To make sure we don't accidentally overflow let's use `strconv.FormatUint([variable], 10)`:

```suggestion
			maxSizeString := strconv.FormatUint(maxSize, 10)
			lastPartitionEndString := strconv.FormatUint(lastPartitionEnd, 10)
			return fmt.Errorf("the MaxSize of %s is not large enough to accomodate defined partitions ending at %s.", maxSizeString, lastPartitionEndString)
```

Comment (PawelWMS) [reply toolkit/tools/imagegen/configuration/disk.go:68]: I'd add `strconv` to the imports, there are just limits to GitHub suggestions.;)

**Review by PawelWMS — APPROVED** [review]
Two small suggestions, looks good. Thanks for adding these checks!

**Review by jslobodzian — APPROVED** [review]

File: toolkit/tools/imagegen/configuration/disk.go [hunk 4]
```diff
@@ -21,13 +23,73 @@ type Disk struct {
 	RawBinaries        []RawBinary        `json:"RawBinaries"`
 }
 
+// checkOverlappingPartitions checks that start and end positions of the defined partitions don't overlap.
+func checkOverlappingePartitions(disk *Disk) (err error) {
+	partIntervals := [][]uint64{}
```
Comment (dmcilvaney) [inline toolkit/tools/imagegen/configuration/disk.go:28]: Very small nit, but this could probably just be `[]*Disk`. No need to change, just took me a sec to wrap my head around the list of lists.

File: toolkit/tools/imagegen/configuration/disk.go [hunk 5]
```diff
@@ -21,13 +23,73 @@ type Disk struct {
...
+	//convert  partition entries to array of [start,end] locations
+	for _, part := range disk.Partitions {
+		partIntervals = append(partIntervals, []uint64{part.Start, part.End})
+	}
+	//sorting paritions by start position
+	sort.Slice(partIntervals, func(i, j int) bool {
+		return partIntervals[i][0] < partIntervals[j][0]
+	})
+	//confirm each partition ends before the next starts
+	for i := 0; i < len(partIntervals)-1; i++ {
```
Comment (dmcilvaney) [inline toolkit/tools/imagegen/configuration/disk.go:38]: this can be `for i,_ := range partIntervals {`. Much more golang style.

Comment (dmcilvaney) [reply toolkit/tools/imagegen/configuration/disk.go:38]: oh, I see you are intending to skip the last entry, could be `range partIntervals[:len(partIntervals) -1 ]` which is a bit of a mess, but still more in the golang style

File: toolkit/tools/imagegen/configuration/disk.go [hunk 6]
```diff
@@ -21,13 +23,73 @@ type Disk struct {
...
+			} else if part.End > lastPartitionEnd {
+				lastPartitionEnd = part.End
+			}
+		}
+		maxSizeString := strconv.FormatUint(maxSize, 10)
+		lastPartitionEndString := strconv.FormatUint(lastPartitionEnd, 10)
+		if maxSize < lastPartitionEnd {
+			return fmt.Errorf("the MaxSize of %s is not large enough to accomodate defined partitions ending at %s.", maxSizeString, lastPartitionEndString)
+		}
+	}
```
Comment (dmcilvaney) [inline toolkit/tools/imagegen/configuration/disk.go:74]: Could we have an `else` here to clearly call out that having a maxsize != 0 paired with a targetdisk.type==realDiskType doesn't make sense either, just to reduce confusion if people are creating a new config.

**Comment by dmcilvaney** [comment]
You can run `make go-test-coverage` to generate `toolkit/out/tools/test_coverage_report.html` which has a nice view of how well our tools are covered by unit tests. I think your tests should cover most of your stuff but its a nice way to double check.

**Comment by dmcilvaney** [comment]
Have a few comments, but generally looks good. I'll approve.

File: toolkit/tools/imagegen/configuration/disk.go (hunk 6 above)
Comment (MaxBrodeurUrbas) [reply toolkit/tools/imagegen/configuration/disk.go:78]: Good point, ive added the relevant warning.
